### PR TITLE
Modify `pr` command to not stop printing at null byte

### LINKED ISSUE #3976: `pr`, unlike `prz` should not stop on zero byte[s]
```
rizin =
[0x0000000]> wx 50505050005050505050
[0x0000000]> px 10
0x0000000 5050 5050 0050 5050 5050                         PPPP.PPPPPP
[0x0000000]> pr
PPPP
[0x0000000]>  pr 10
PPPPPPPPP
```
It should return whole block (on `pr` without arguments) or requested size (with argument).

See `librz/core/cmd/cmd_print.c`

### PR DESCRIPTION
**Pull request by HN026**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/3976

### CODE REVIEW

**Comment by rickhg12hs** [comment]
#3976 
Shouldn't `pr` just print raw bytes, regardless of value?  Isn't `prz` for printing null-terminating strings?

File: librz/core/cmd/cmd_print.c [hunk 1]
```diff
@@ -6969,7 +6969,13 @@ static void printraw(RzCore *core, int len) {
...
 	}
-	if (rz_io_read_at(core->io, core->offset, data, len)) {
+	if(rz_io_read_at(core->io, core->offset, data, len)){
+		for (int i = 0; i<len; i++){
+			if(data[i]=='\0'){
+				len = i;
+				break;
+			}
+		}
 		rz_print_raw(core->print, core->offset, data, len);
```
Comment (wargio) [inline librz/core/cmd/cmd_print.c]: no. `prz` (print raw zero) is the one that stops at the first `\0` byte, meanwhile `pr` (print raw) should keep printing till the end of the buffer

Comment (wargio) [reply librz/core/cmd/cmd_print.c]: Check how `rz_print_raw` works and modify it to allow to stop the printing when a zero or not. then modify all the calls to this method to either print till the end of the buffer or till the first zero.

Comment (rickhg12hs) [reply librz/core/cmd/cmd_print.c]: I wouldn't trust my programming skills, but shouldn't:

https://github.com/rizinorg/rizin/blob/b9851b7c24ab5d3cc2d0192d4b8f8859cad2999c/librz/core/cmd/cmd_print.c#L6979-L6986

really be:
``` c
RZ_IPI RzCmdStatus rz_cmd_print_raw_handler(RzCore *core, int argc, const char **argv) {
        int len = argc > 1 ? rz_num_math(core->num, argv[1]) : core->blocksize;
        if (len < 0) {
                return RZ_CMD_STATUS_ERROR;
        }
        printraw(core, len);
        return RZ_CMD_STATUS_OK;
}
```
I.e., don't try to determine any string length.  Use the block size.  (I don't know what sort of downstream checking is done, or even if it's necessary, but I wonder what would happen if `len` is so large that something/somewhere points off into nothingness, or something/somewhere that it shouldn't.)

Comment (wargio) [reply librz/core/cmd/cmd_print.c]: definitely you cannot call `strlen`

Comment (HN026) [reply librz/core/cmd/cmd_print.c]: Suree, Will do it

Comment (HN026) [reply librz/core/cmd/cmd_print.c]: Basically, a rough idea of what we can do is in our rz_print_raw function we can add another parameter that basically tells us if we have to stop at the first \0 or not, then modify all the other functions that are calling this method.

File: librz/core/cmd/cmd_print.c [hunk 2]
```diff
@@ -6970,14 +6970,26 @@ static void printraw(RzCore *core, int len) {
 		return;
 	}
 	if (rz_io_read_at(core->io, core->offset, data, len)) {
-		rz_print_raw(core->print, core->offset, data, len);
+		rz_print_raw(core->print, core->offset, data, len, false);
+	}
+	free(data);
```
Comment (wargio) [inline librz/core/cmd/cmd_print.c:6975]: nice catch

File: librz/core/cmd/cmd_print.c [hunk 3]
```diff
@@ -6970,14 +6970,26 @@ static void printraw(RzCore *core, int len) {
...
+	core->cons->newline = core->cmd_in_backticks ? false : true;
+}
+
+static void printrawzero(RzCore *core, int len) {
```
Comment (wargio) [inline librz/core/cmd/cmd_print.c]: instead of duplicating, add an argument to this and pass it to rz_print_raw.

File: librz/util/print.c [hunk 4]
```diff
@@ -1245,8 +1245,17 @@ RZ_API void rz_print_bytes(RzPrint *p, const ut8 *buf, int len, const char *fmt)
...
-RZ_API void rz_print_raw(RzPrint *p, ut64 addr, const ut8 *buf, int len) {
-	p->write(buf, len);
+RZ_API void rz_print_raw(RzPrint *p, ut64 addr, const ut8 *buf, int len, bool stop_at_null) {
+	if (stop_at_null) {
+		for (int i = 0; i < len; i++) {
+			if (buf[i] == '\0') {
+				break;
+			}
+			p->write(&buf[i], 1);
+		}
```
Comment (wargio) [inline librz/util/print.c]: this is ok, but is much more faster to call `rz_str_nlen` to get the actual len and then call the write.

File: librz/util/print.c [hunk 5]
```diff
@@ -1245,8 +1245,17 @@ RZ_API void rz_print_bytes(RzPrint *p, const ut8 *buf, int len, const char *fmt)
...
+	} else {
+		p->write(buf, len);
+	}
```
Comment (wargio) [inline librz/util/print.c]:
```suggestion
	if (stop_at_null) {
		len = rz_str_nlen(buf, len);
	}
	p->write(buf, len);
```

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add a test

File: test/db/cmd/print [hunk 6]
```diff
@@ -246,3 +246,31 @@ struct elf_header {
...
+4000 0000 0000 0000 4000 0000 0000 0000  @.......@.......
+6802 0000
+EOF 
+RUN
+
+NAME=prz cmd with offset 0
+FILE=bins/elf/ls
+CMDS=<<EOF
+s 0
+prz 100
```
Comment (wargio) [inline test/db/cmd/print]:
```suggestion
prz 100 | rz-ax -S
```

File: test/db/cmd/print [hunk 7]
```diff
@@ -246,3 +246,31 @@ struct elf_header {
...
+EOF
+EXPECT=<<EOF
+ELF
```
Comment (wargio) [inline test/db/cmd/print]: 7f454c46020101

**Review by wargio — CHANGES_REQUESTED** [review]
just modify the test and i think we can merge.

**Comment by wargio** [comment]
also format the code.

**Review by rickhg12hs — COMMENTED** [review]
In light of [MatterMost message: "Why does `pr` stop there?"](https://im.rizin.re/rizinorg/pl/hd9gk9n95prffetcap5itogd6y), I'm wondering if a better test would be `pr @! 100` rather than `pr 100`.  `pr 100` works without this PR.

**Review by wargio — APPROVED** [review]

**Comment by HN026** [comment]
@wargio what do you think about this

File: librz/core/cmd/cmd_print.c [hunk 8]
```diff
@@ -7028,10 +7028,10 @@ RZ_IPI RzCmdStatus rz_cmd_print_raw_printable_handler(RzCore *core, int argc, co
 }
 
 RZ_IPI RzCmdStatus rz_cmd_print_raw_string_handler(RzCore *core, int argc, const char **argv) {
-	int len = argc > 1 ? rz_num_math(core->num, argv[1]) : strlen((const char *)core->block);
+	int len = argc > 1 ? rz_num_math(core->num, argv[1]) : core->blocksize;
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:7034]: i'm not convinced we want to limit a "string-based command" with blocksize. It seems a bit counter intuitive. @wargio WDYT?

Comment (wargio) [reply librz/core/cmd/cmd_print.c:7034]: we said before that due UX, we want to have these commands to accept a buffer size and when not supplied the current block size.

Comment (rickhg12hs) [reply librz/core/cmd/cmd_print.c:7034]: Would `pr @! some_number` allow a user to specify a max?

Comment (XVilka) [reply librz/core/cmd/cmd_print.c:7034]: Agree with @wargio

Comment (wargio) [reply librz/core/cmd/cmd_print.c:7034]: > Would `pr @! some_number` allow a user to specify a max?

yes also. this is why the command also uses the blocksize variable

Comment (HN026) [reply librz/core/cmd/cmd_print.c:7034]: Should I change the test to pr @! 100? @wargio

**Comment by wargio** [comment]
these changes broke a test. try to understand why and fix the issue. the test is correct, so you don't need to modify it

**Comment by HN026** [comment]
@wargio  the changes we made regarding the **core->blocksize** to be sure you're talking about those ones?

File: librz/core/cmd/cmd_print.c (hunk 8 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:7034]: no need

**Comment by rickhg12hs** [comment]
How _"raw"_ is `printraw` and `rz_print_raw` if they optionally check for `stop_at_null`?  To keep the `...raw` routines _"raw"_, shouldn't they just use the start and `len`?  ... and have `len` properly determined before calling the `...raw` routine?

**Comment by wargio** [comment]
> How _"raw"_ is `printraw` and `rz_print_raw` if they optionally check for `stop_at_null`? To keep the `...raw` routines _"raw"_, shouldn't they just use the start and `len`? ... and have `len` properly determined before calling the `...raw` routine?

good point. that is true, would make more sense to do this before calling `rz_print_raw` inside the `printraw` function.
@HN026 move the changes in `rz_print_raw` inside `printraw` and revert back the function definitions of `rz_print_raw`.

**Comment by rickhg12hs** [comment]
> > How _"raw"_ is `printraw` and `rz_print_raw` if they optionally check for `stop_at_null`? To keep the `...raw` routines _"raw"_, shouldn't they just use the start and `len`? ... and have `len` properly determined before calling the `...raw` routine?
> 
> good point. that is true, would make more sense to do this before calling `rz_print_raw` inside the `printraw` function. 

Is it even necessary/desired within `printraw` to have `stop_at_null` functionality?

**Comment by wargio** [comment]
well, we avoid code duplication.

**Comment by rickhg12hs** [comment]
> well, we avoid code duplication.

Ah, OK.  I thought, perhaps mistakenly, that https://github.com/rizinorg/rizin/blob/c18266311fc5647032d069e0596aad471a72fe82/librz/core/cmd/cmd_print.c#L6979
exclusively handled `pr` and any `len` fixing could be done there.

I should return to _user_ mode and avoid _developer_ mode.  😊

**Review by wargio — APPROVED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
Fix tests:
```
[XX] db/cmd/print pr cmd with offset 0 
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 's 0
pr 100
' bins/elf/ls
-- stdout
--- expected
+++ actual
@@ -1,7 +1,1 @@
-7f45 4c46 0201 0100 0000 0000 0000 0000  .ELF............
-0300 3e00 0100 0000 e05a 0000 0000 0000  ..>......Z......
-4000 0000 0000 0000 6813 0200 0000 0000  @.......h.......
-0000 0000 4000 3800 0b00 4000 1900 1800  ....@.8...@.....
-0600 0000 0400 0000 4000 0000 0000 0000  ........@.......
-4000 0000 0000 0000 4000 0000 0000 0000  @.......@.......
-6802 0000
+ELF

-- stderr

[XX] db/archos/linux-any/cmd_i utf8 (possibly malformed) filenames with non-printable chars
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'cd `e dir.tmp`
e io.cache=true
(show_fname fname; wz ${fname}; pr `$size` > `prz` ; o `prz`; i~file; ij; rm `prz`)
.(show_fname A\x1b¢\302\200€𝄞\363\240\201\201\\.bin)
.(show_fname B\x1b¢\302\200€𝄞\363\240\201\201\\.bin)
.(show_fname B\x1b¢\302\200\201€𝄞\363\240\201\201\\.bin)
$size=%v 1
.(show_fname €\\.bin)
.(show_fname \xf0\x82\x82\xac\\.bin)
' =
-- stdout
--- expected
+++ actual
@@ --1,11 +-1,11 @@
 file    A\e¢€𝄞󠁁\.bin
-{"core":{"file":"A\u001b¢\u0080€𝄞\udb40\udc41\\.bin","fd":4,"size":22,"humansz":"22","iorw":true,"mode":"r-x","block":256,"format":"any"}}
+{"core":{"file":"A\u001b¢\u0080€𝄞\udb40\udc41\\.bin","fd":4,"size":256,"humansz":"256","iorw":true,"mode":"r-x","block":256,"format":"any"}}
 file    B\e¢€𝄞󠁁\.bin
-{"core":{"file":"B\u001b¢\u0080€𝄞\udb40\udc41\\.bin","fd":5,"size":22,"humansz":"22","iorw":true,"mode":"r-x","block":256,"format":"any"}}
+{"core":{"file":"B\u001b¢\u0080€𝄞\udb40\udc41\\.bin","fd":5,"size":256,"humansz":"256","iorw":true,"mode":"r-x","block":256,"format":"any"}}
 file    B\e¢\x81€𝄞󠁁\.bin
-{"core":{"file":[66,27,194,162,194,128,129,226,130,172,240,157,132,158,243,160,129,129,92,46,98,105,110],"fd":6,"size":23,"humansz":"23","iorw":true,"mode":"r-x","block":256,"format":"any"}}
+{"core":{"file":[66,27,194,162,194,128,129,226,130,172,240,157,132,158,243,160,129,129,92,46,98,105,110],"fd":6,"size":256,"humansz":"256","iorw":true,"mode":"r-x","block":256,"format":"any"}}
 file    €\.bin
 {"core":{"file":"€\\.bin","fd":7,"size":1,"humansz":"1","iorw":true,"mode":"r-x","block":256,"format":"any"}}
 file    \xf0\x82\x82\xac\.bin

-- stderr

```

**Comment by wargio** [comment]
change the test to use `prz` instead of `pr`, as shown below.
- Old:
```
(show_fname fname; wz ${fname}; pr `$size` > `prz` ; o `prz`; i~file; ij; rm `prz`)
```
 - New 
```
(show_fname fname; wz ${fname}; prz `$size` > `prz` ; o `prz`; i~file; ij; rm `prz`)
```
Also in the test for `pr cmd with offset 0 ` add a pipe to `rz-ax` like in the other example.

**Comment by wargio** [comment]
Please fix the broken test.

**Comment by HN026** [comment]
> Please fix the broken test.

On it

**Comment by wargio** [comment]
```
[XX] db/cmd/print pr cmd with offset 0 
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 's 0
pr 100 | rz-ax -S
' bins/elf/ls
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-f30f1efa31ed4989d15e4889e24883e4f050544c8d05660c0100488d0def0b0100488d3d68e5ffffff150ac30100f490488d3d51c70100488d054ac701004839f87415488b0596c101004885c07409ffe00f1f8000000000c30f1f8000000000488d3d21
+7f454c4602010100000000000000000003003e0001000000e05a0000000000004000000000000000681302000000000000000000400038000b00400019001800060000000400000040000000000000004000000000000000400000000000000068020000

-- stderr
```

**Comment by wargio** [comment]
fix also `db/archos/darwin-x64/cmd_i utf8 (possibly malformed) filenames with non-printable chars`

File: test/db/archos/darwin-x64/cmd_i [hunk 9]
```diff
@@ -4,7 +4,7 @@ FILE==
 CMDS=<<EOF
 cd `e dir.tmp`
 e io.cache=true
-(show_fname fname; wz ${fname}; pr `$b` > `prz`; o `prz`; i~file; ij; rm `prz`)
+(show_fname fname; wz ${fname}; prz`$b` > `prz`; o `prz`; i~file; ij; rm `prz`)
```
Comment (wargio) [inline test/db/archos/darwin-x64/cmd_i]:
```suggestion
(show_fname fname; wz ${fname}; prz `$b` > `prz`; o `prz`; i~file; ij; rm `prz`)
```

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]
lgtm.

**Pull request merged**